File: algo/MergeTwoSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2)
    {
        ListNode *res = new ListNode(0);
        ListNode *p = res;

        while (list1 != NULL && list2 != NULL)
        {
            if (list1->val < list2->val)
            {
                p->next = new ListNode(list1->val);
                list1 = list1->next;
            }
            else
            {
                p->next = new ListNode(list2->val);
                list2 = list2->next;
            }
            p = p->next;
        }
        if (list1 != NULL)
        {
            p->next = list1;
        }
        else
        {
            p->next = list2;
        }
        return res->next;
    }
};

File: algo/MergeTwoSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class MergeTwoSortedLists {
    // 21. Merge Two Sorted Lists

    // Approach 1: Iterative
    // Time Complexity: O(Max(N,M))
    // Space Complexity: O(Max(N,M))
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode res = new ListNode();
        ListNode p = res;

        ListNode p1 = list1;
        ListNode p2 = list2;
        while (p1 != null && p2 != null) {
            if (p1.val < p2.val) {
                p.next = p1;
                p1 = p1.next;
            } else {
                p.next = p2;
                p2 = p2.next;
            }
            p = p.next;
        }
        if (p1 != null) {
            p.next = p1;
        } else {
            p.next = p2;
        }
        return res.next;
    }
}